utils/include/detray/simulation/random_scatterer.hpp

```diff
@@ -42,23 +65,79 @@ struct random_scatterer : actor {
         void set_seed(const uint_fast64_t sd) { generator.seed(sd); }
     };
 
+    /// Material store visitor
+    struct kernel {
+
+        using scalar_type = typename interaction_type::scalar_type;
+        using state = typename random_scatterer::state;
+
+        template <typename material_group_t, typename index_t,
+                  typename surface_t>
+        DETRAY_HOST_DEVICE inline void operator()(
+            const material_group_t& material_group,
+            const index_t& material_range,
+            const intersection2D<surface_t, transform3_type>& is, state& s,
+            const bound_track_parameters<transform3_type>& bound_params) const {
+
+            const scalar qop = bound_params.qop();
+            const scalar charge = bound_params.charge();
+
+            for (const auto& mat :
+                 detray::ranges::subrange(material_group, material_range)) {
+
+                // Energy Loss
+                if (s.do_energy_loss) {
+                    s.e_loss_mpv =
+                        interaction_type().compute_energy_loss_landau(
+                            is, mat, s.pdg, s.mass, qop, charge);
+
+                    s.e_loss_sigma =
+                        interaction_type().compute_energy_loss_landau_sigma(
+                            is, mat, s.pdg, s.mass, qop, charge);
+                }
+
+                // Covariance update
+                if (s.do_multiple_scattering) {
+                    // @todo: use momentum before or after energy loss in
+                    // backward mode?
+                    s.projected_scattering_angle =
+                        interaction_type().compute_multiple_scattering_theta0(
+                            is, mat, s.pdg, s.mass, qop, charge);
+                }
+            }
+        }
+    };
+
     /// Observes a material interactor state @param interactor_state
-    template <typename interactor_state_t, typename propagator_state_t>
+    template <typename propagator_state_t>
     DETRAY_HOST inline void operator()(state& simulator_state,
-                                       interactor_state_t& interactor_state,
                                        propagator_state_t& prop_state) const {
 
         auto& navigation = prop_state._navigation;
 
         if (navigation.is_on_module()) {
 
             auto& stepping = prop_state._stepping;
+            auto& bound_params = stepping._bound_params;
+            const auto* det = navigation.detector();
+            const auto is = *navigation.current();
```

Comment:
You might want to catch this as a reference explicitly

Reply:
ok